// @material/mwc-notched-outline/mwc-notched-outline-base.js
import { __decorate } from '../../tslib/tslib.es6.js';
import { BaseElement } from '../mwc-base/base-element.js';
import { MDCNotchedOutlineFoundation } from '../notched-outline/foundation.js';
import '../../@lit/reactive-element/reactive-element.js';
import { html as y } from '../../lit-html/lit-html.js';
import '../../lit-element/lit-element.js';
import { property as e } from '../../@lit/reactive-element/decorators/property.js';
import { query as i } from '../../@lit/reactive-element/decorators/query.js';
import '../../@lit/reactive-element/decorators/query-assigned-elements.js';
import { classMap as o } from '../../lit-html/directives/class-map.js';

/**
 * @license
 * Copyright 2019 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
class NotchedOutlineBase extends BaseElement {
    constructor() {
        super(...arguments);
        this.mdcFoundationClass = MDCNotchedOutlineFoundation;
        this.width = 0;
        this.open = false;
        this.lastOpen = this.open;
    }
    createAdapter() {
        return {
            addClass: (className) => this.mdcRoot.classList.add(className),
            removeClass: (className) => this.mdcRoot.classList.remove(className),
            setNotchWidthProperty: (width) => this.notchElement.style.setProperty('width', `${width}px`),
            removeNotchWidthProperty: () => this.notchElement.style.removeProperty('width'),
        };
    }
    openOrClose(shouldOpen, width) {
        if (!this.mdcFoundation) {
            return;
        }
        if (shouldOpen && width !== undefined) {
            this.mdcFoundation.notch(width);
        }
        else {
            this.mdcFoundation.closeNotch();
        }
    }
    render() {
        this.openOrClose(this.open, this.width);
        const classes = o({
            'mdc-notched-outline--notched': this.open,
        });
        return y `
      <span class="mdc-notched-outline ${classes}">
        <span class="mdc-notched-outline__leading"></span>
        <span class="mdc-notched-outline__notch">
          <slot></slot>
        </span>
        <span class="mdc-notched-outline__trailing"></span>
      </span>`;
    }
}
__decorate([
    i('.mdc-notched-outline')
], NotchedOutlineBase.prototype, "mdcRoot", void 0);
__decorate([
    e({ type: Number })
], NotchedOutlineBase.prototype, "width", void 0);
__decorate([
    e({ type: Boolean, reflect: true })
], NotchedOutlineBase.prototype, "open", void 0);
__decorate([
    i('.mdc-notched-outline__notch')
], NotchedOutlineBase.prototype, "notchElement", void 0);

export { NotchedOutlineBase };
